File: Mage/AttachmentViewCoordinator.m
//
//  AttachmentViewCoordinator.m
//  MAGE
//
//  Created by Daniel Barela on 3/10/20.
//  Copyright © 2020 National Geospatial Intelligence Agency. All rights reserved.
//

#import "AttachmentViewCoordinator.h"
#import "DataConnectionUtilities.h"
#import "AttachmentViewController.h"
#import "FadeTransitionSegue.h"
#import "AVFoundation/AVFoundation.h"
#import <MediaPlayer/MediaPlayer.h>
#import <AVFoundation/AVFoundation.h>
#import <AVKit/AVKit.h>
#import <MobileCoreServices/MobileCoreServices.h>
#import "StoredPassword.h"
#import "MAGE-Swift.h"

@interface AttachmentViewCoordinator() <NSURLConnectionDataDelegate, AVAssetResourceLoaderDelegate, UINavigationControllerDelegate, NavigationControllerObserverDelegate>

@property (strong, nonatomic) UINavigationController *rootViewController;
@property (strong, nonatomic) UINavigationController *navigationController;
@property (strong, nonatomic) NavigationControllerObserver *navigationControllerObserver;
@property (strong, nonatomic) id<AttachmentViewDelegate> delegate;
@property (strong, nonatomic) Attachment *attachment;
@property (strong, nonatomic) NSMutableArray *pendingRequests;
@property (strong, nonatomic) AVPlayer *player;
@property (strong, nonatomic) NSURLConnection *connection;
@property (nonatomic, strong) NSHTTPURLResponse *response;
@property (strong, nonatomic) AVPlayerViewController *playerViewController;
@property (strong, nonatomic) NSString *tempFile;
@property (nonatomic, assign) NSUInteger fullAudioDataLength;
@property (nonatomic, strong) NSURL *urlToLoad;

@end

@implementation AttachmentViewCoordinator

- (instancetype) initWithNavigationController: (UINavigationController *) rootViewController andDelegate: (id<AttachmentViewDelegate>) delegate andAttachment: (Attachment *) attachment {
    self = [super init];
    if (!self) return nil;
    
    _rootViewController = rootViewController;
    
    _delegate = delegate;
    _attachment = attachment;
    
    self.tempFile = [NSTemporaryDirectory() stringByAppendingPathComponent:[self.attachment.url lastPathComponent]];
    self.navigationController = rootViewController;
    self.navigationControllerObserver = [[NavigationControllerObserver alloc] initWithNavigationController:self.navigationController];
    
    return self;
}

- (void) start {
//    self.navigationController.navigationItem.backBarButtonItem.title = self.rootViewController.title;
//    [self.navigationController setModalPresentationStyle:UIModalPresentationFullScreen];
//    [self.rootViewController presentViewController:self.navigationController animated:YES completion:nil];
//    [self.rootViewController pushViewController: self.navigationController animated:YES];
    if (![DataConnectionUtilities shouldFetchAttachments] && !self.attachment.localPath && ![[NSFileManager defaultManager] fileExistsAtPath:self.tempFile]) {
        AskToDownloadViewController *vc = [[AskToDownloadViewController alloc] initWithAttachment:self.attachment delegate:self];
        [_navigationController pushViewController:vc animated:YES];
        [self.navigationControllerObserver observePopTransitionOf:vc delegate:self];
        return;
    } else {
        if ([self.attachment.contentType hasPrefix:@"image"]) {
            ImageAttachmentViewController *ac = [[ImageAttachmentViewController alloc] initWithAttachment:self.attachment];
            ac.view.backgroundColor = [UIColor blackColor];
            [_navigationController pushViewController:ac animated:YES];
        } else if ([self.attachment.contentType hasPrefix:@"video"] || [self.attachment.contentType hasPrefix:@"audio"]) {
            [self downloadAudioVideo];
        }
    }
    
}

- (void) navigationControllerObserver:(NavigationControllerObserver *)observer didObservePopTransitionFor:(UIViewController *)viewController {
    // this gets called on the back button
    NSLog(@"pop %@", viewController);
    [self.delegate doneViewing:self];
}

- (void) navigationController:(UINavigationController *)navigationController didShowViewController:(UIViewController *)viewController animated:(BOOL)animated {
    NSLog(@"There are %ld view controllers", [self.navigationController.viewControllers count]);
    if ([navigationController.viewControllers count] == 0) {
        [self.delegate doneViewing:self];
    }
}

- (void) downloadAttachment {
    // proceed to the attachment downloading view
    NSLog(@"Download the attachment");
    [FadeTransitionSegue addFadeTransitionToView:self.navigationController.view];

    [_navigationController popViewControllerAnimated:NO];
    
    if ([self.attachment.contentType hasPrefix:@"image"]) {
        ImageAttachmentViewController *ac = [[ImageAttachmentViewController alloc] initWithAttachment:self.attachment];
        [_navigationController pushViewController:ac animated:NO];
    } else if ([self.attachment.contentType hasPrefix:@"video"] || [self.attachment.contentType hasPrefix:@"audio"]) {
        [self downloadAudioVideo];
    }
}

- (void) downloadAudioVideo {
    if ([[NSFileManager defaultManager] fileExistsAtPath:self.attachment.localPath]) {
        // save the local path
        NSLog(@"playing locally");
        return [self playMediaType:self.attachment.contentType FromDocumentsFolder:self.attachment.localPath];
    }
    NSString *urlString = [NSString stringWithFormat: @"%@?access_token=%@", self.attachment.url, [StoredPassword retrieveStoredToken]];
    NSURL *url = [NSURL URLWithString:urlString];
    self.urlToLoad = url;
    NSURLComponents *components = [NSURLComponents componentsWithURL:url resolvingAgainstBaseURL:NO];
    components.scheme = @"streaming";
    AVURLAsset *asset = [AVURLAsset URLAssetWithURL:components.URL options:nil];
    [asset.resourceLoader setDelegate:self queue:dispatch_get_main_queue()];
    // This tracks all pending AVAssetResourceLoadingRequest objects we have not fulfilled yet
    self.pendingRequests = [NSMutableArray array];
    
    AVPlayerItem *playerItem = [AVPlayerItem playerItemWithAsset:asset];
    self.player = [[AVPlayer alloc] initWithPlayerItem:playerItem];
    [playerItem addObserver:self forKeyPath:@"status" options:NSKeyValueObservingOptionNew context:NULL];
    
    self.playerViewController = [[AVPlayerViewController alloc] init];
    self.playerViewController.player = self.player;
    [self.playerViewController.view setAutoresizingMask:UIViewAutoresizingFlexibleHeight | UIViewAutoresizingFlexibleWidth];
    [_navigationController pushViewController:self.playerViewController animated:NO];
    [self.navigationControllerObserver observePopTransitionOf:self.playerViewController delegate:self];
}

-(void) playMediaType: (NSString *) type FromDocumentsFolder:(NSString *) fromPath {
    NSURL *url = [NSURL fileURLWithPath:fromPath];
    NSLog(@"Playing %@", url);
    self.urlToLoad = url;
//    self.playerViewController = [[AVPlayerViewController alloc] init];
//    self.playerViewController.player = [AVPlayer playerWithURL:url];
//
//    [self.playerViewController.view setAutoresizingMask:UIViewAutoresizingFlexibleHeight | UIViewAutoresizingFlexibleWidth];
//    [_navigationController pushViewController:self.playerViewController animated:NO];
//    [self.navigationControllerObserver observePopTransitionOf:self.playerViewController delegate:self];
//
//    [self.playerViewController.player play];
    NSURLComponents *components = [NSURLComponents componentsWithURL:url resolvingAgainstBaseURL:NO];
    components.scheme = @"streaming";
    AVURLAsset *asset = [AVURLAsset URLAssetWithURL:components.URL options:nil];
    [asset.resourceLoader setDelegate:self queue:dispatch_get_main_queue()];
    // This tracks all pending AVAssetResourceLoadingRequest objects we have not fulfilled yet
    self.pendingRequests = [NSMutableArray array];
    
    AVPlayerItem *playerItem = [AVPlayerItem playerItemWithAsset:asset];
    self.player = [[AVPlayer alloc] initWithPlayerItem:playerItem];
    [playerItem addObserver:self forKeyPath:@"status" options:NSKeyValueObservingOptionNew context:NULL];
    
    self.playerViewController = [[AVPlayerViewController alloc] init];
    self.playerViewController.player = self.player;
    [self.playerViewController.view setAutoresizingMask:UIViewAutoresizingFlexibleHeight | UIViewAutoresizingFlexibleWidth];
    [_navigationController pushViewController:self.playerViewController animated:NO];
    [self.navigationControllerObserver observePopTransitionOf:self.playerViewController delegate:self];
}

#pragma mark - NSURLConnection delegate

- (void)connection:(NSURLConnection *)connection didReceiveResponse:(NSURLResponse *)response
{
    self.fullAudioDataLength = 0;
    self.response = (NSHTTPURLResponse *)response;
    
    [self processPendingRequests];
}

- (void)connection:(NSURLConnection *)connection didReceiveData:(NSData *)data
{
    self.fullAudioDataLength += data.length;
    [self appendDataToTempFile: data];
    
    [self processPendingRequests];
}

- (void)appendDataToTempFile:(NSData *)data
{
    if(![[NSFileManager defaultManager] fileExistsAtPath:self.tempFile]) {
        [data writeToFile:self.tempFile atomically:YES];
//        [data writeToURL:url atomically:YES];
//        [FileUtils saveData:data toURL:self.tempURL];
    } else {
        NSFileHandle *fileHandle = [NSFileHandle fileHandleForWritingAtPath:self.tempFile];
        [fileHandle seekToEndOfFile];
        [fileHandle writeData:data];
    }
}

- (void)connectionDidFinishLoading:(NSURLConnection *)connection
{
    NSLog(@"finished loading the media file");
    // save it and set the local path of the attachment
    [MagicalRecord saveWithBlockAndWait:^(NSManagedObjectContext *localContext) {
        Attachment *localAttachment = [self.attachment MR_inContext:localContext];
        localAttachment.localPath = self.tempFile;
    }];
    [self processPendingRequests];
}

#pragma mark - AVURLAsset resource loading

- (void)processPendingRequests
{
    NSMutableArray *requestsCompleted = [NSMutableArray array];
    
    for (AVAssetResourceLoadingRequest *loadingRequest in self.pendingRequests)
    {
        [self fillInContentInformation:loadingRequest.contentInformationRequest];
        
        BOOL didRespondCompletely = [self respondWithDataForRequest:loadingRequest.dataRequest];
        
        if (didRespondCompletely)
        {
            [requestsCompleted addObject:loadingRequest];
            
            [loadingRequest finishLoading];
        }
    }
    
    [self.pendingRequests removeObjectsInArray:requestsCompleted];
}

- (void)fillInContentInformation:(AVAssetResourceLoadingContentInformationRequest *)contentInformationRequest
{
    if (contentInformationRequest == nil || self.response == nil)
    {
        return;
    }
    
    NSString *mimeType = [self.response MIMEType];
    CFStringRef contentType = UTTypeCreatePreferredIdentifierForTag(kUTTagClassMIMEType, (__bridge CFStringRef)(mimeType), NULL);
    
    contentInformationRequest.byteRangeAccessSupported = YES;
    contentInformationRequest.contentType = CFBridgingRelease(contentType);
    contentInformationRequest.contentLength = [self.response expectedContentLength];
}

- (BOOL)respondWithDataForRequest:(AVAssetResourceLoadingDataRequest *)dataRequest
{
    long long startOffset = dataRequest.requestedOffset;
    if (dataRequest.currentOffset != 0)
    {
        startOffset = dataRequest.currentOffset;
    }
    
    // Don't have any data at all for this request
    if (self.fullAudioDataLength < startOffset)
    {
        return NO;
    }
    
    // This is the total data we have from startOffset to whatever has been downloaded so far
    NSUInteger unreadBytes = self.fullAudioDataLength - (NSUInteger)startOffset;
    
    // Respond with whatever is available if we can't satisfy the request fully yet
    NSUInteger numberOfBytesToRespondWith = MIN((NSUInteger)dataRequest.requestedLength, unreadBytes);
    
    NSRange range = NSMakeRange((NSUInteger)startOffset, numberOfBytesToRespondWith);
    NSData *subData = [self dataFromFileInRange:range];
    [dataRequest respondWithData:subData];
    
    long long endOffset = startOffset + dataRequest.requestedLength;
    BOOL didRespondFully = self.fullAudioDataLength >= endOffset;
    
    return didRespondFully;
}

- (NSData *)dataFromFileInRange:(NSRange)range
{
    NSFileHandle *fileHandle = [NSFileHandle fileHandleForReadingAtPath:self.tempFile];;
    [fileHandle seekToFileOffset:range.location];
    return [fileHandle readDataOfLength:range.length];
}


- (BOOL)resourceLoader:(AVAssetResourceLoader *)resourceLoader shouldWaitForLoadingOfRequestedResource:(AVAssetResourceLoadingRequest *)loadingRequest
{
    if (self.connection == nil)
    {
        NSURL *interceptedURL = [loadingRequest.request URL];
        NSURLComponents *actualURLComponents = [[NSURLComponents alloc] initWithURL:interceptedURL resolvingAgainstBaseURL:NO];
        NSURLComponents *loadingComponents = [[NSURLComponents alloc] initWithURL:self.urlToLoad resolvingAgainstBaseURL:NO];
        actualURLComponents.scheme = loadingComponents.scheme;
        
        NSURLRequest *request = [NSURLRequest requestWithURL:[actualURLComponents URL]];
        self.connection = [[NSURLConnection alloc] initWithRequest:request delegate:self startImmediately:NO];
        [self.connection setDelegateQueue:[NSOperationQueue mainQueue]];
        
        [self.connection start];
    }
    
    [self.pendingRequests addObject:loadingRequest];
    [self processPendingRequests];
    
    return YES;
}

- (void)resourceLoader:(AVAssetResourceLoader *)resourceLoader didCancelLoadingRequest:(AVAssetResourceLoadingRequest *)loadingRequest
{
    [self.pendingRequests removeObject:loadingRequest];
}

#pragma KVO

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context
{
    if (self.player.currentItem.status == AVPlayerItemStatusReadyToPlay)
    {
        [self.player play];
    }
}
@end

File: Mage/AttachmentViewCoordinator.swift
//
//  AttachmentViewCoordinator.m
//  MAGE
//
//  Created by Daniel Barela on 3/10/20.
//  Copyright © 2020 National Geospatial Intelligence Agency. All rights reserved.
//

import UIKit
import AVKit
import MagicalRecord
import Kingfisher
import QuickLook

@objc protocol AttachmentViewDelegate {
    @objc func doneViewing(coordinator: NSObject)
}

@objc class AttachmentViewCoordinator: NSObject, NavigationControllerObserverDelegate, AskToDownloadDelegate {
    @Injected(\.attachmentRepository) var attachmentRepository: AttachmentRepository
    @Injected(\.attachmentLocalDataSource) private var localSD: AttachmentLocalDataSource

    var scheme: MDCContainerScheming?

    var attachment: AttachmentModel?
    weak var delegate: AttachmentViewDelegate?
    var rootViewController: UINavigationController
    var navigationControllerObserver: NavigationControllerObserver
    var tempFile: String?
    var contentType: String?

    var mediaPreviewController: MediaPreviewController?
    var imageViewController: ImageAttachmentViewController?
    var observer: NSKeyValueObservation?

    var urlToLoad: URL?
    var fullAudioDataLength: Int = 0

    var hasPushedViewController: Bool = false
    var ignoreNextDelegateCall: Bool = false

    // MARK: - Helpers

    private func healedLocalURL(for att: AttachmentModel?) -> URL? {
        guard let att else { return nil }
        return AttachmentPath.localURL(fromStored: att.localPath, fileName: att.name)
    }

    @inline(__always)
    private func onMain(_ work: @escaping () -> Void) {
        if Thread.isMainThread { work() } else { DispatchQueue.main.async(execute: work) }
    }

    // MARK: - Init

    @objc public init(rootViewController: UINavigationController, attachment: AttachmentModel, delegate: AttachmentViewDelegate?, scheme: MDCContainerScheming?, navigationControllerObserver: NavigationControllerObserver? = nil) {
        self.rootViewController = rootViewController
        self.attachment = attachment
        self.delegate = delegate
        self.scheme = scheme
        contentType = attachment.contentType

        if let attachmentUrl = self.attachment?.url {
            self.tempFile = NSTemporaryDirectory() + (URL(string: attachmentUrl)?.lastPathComponent ?? "tempfile")
        } else {
            self.tempFile = NSTemporaryDirectory() + "tempfile"
        }
        self.navigationControllerObserver = navigationControllerObserver ?? NavigationControllerObserver(navigationController: self.rootViewController)
        super.init()
    }

    @objc public init(rootViewController: UINavigationController, url: URL, contentType: String, delegate: AttachmentViewDelegate?, scheme: MDCContainerScheming?, navigationControllerObserver: NavigationControllerObserver? = nil) {
        self.rootViewController = rootViewController
        self.urlToLoad = url
        self.delegate = delegate
        self.scheme = scheme
        self.contentType = contentType
        self.tempFile = NSTemporaryDirectory() + url.lastPathComponent

        self.navigationControllerObserver = navigationControllerObserver ?? NavigationControllerObserver(navigationController: self.rootViewController)
        super.init()
    }

    // MARK: - Flow

    @objc public func start() { start(true) }

    @objc public func start(_ animated: Bool = true) {
        onMain {
            if UIDevice.current.userInterfaceIdiom == .pad {
                let navigationController = UINavigationController()
                navigationController.view.backgroundColor = self.scheme?.colorScheme.surfaceColor
                self.rootViewController.present(navigationController, animated: animated, completion: nil)
                self.rootViewController = navigationController
            }
            if let theAttachment = self.attachment {
                // Prefer healed local URL over raw string checks
                if self.healedLocalURL(for: theAttachment) != nil || DataConnectionUtilities.shouldFetchAttachments() {
                    self.showAttachment(animated: animated)
                } else {
                    let vc = AskToDownloadViewController(attachment: theAttachment, delegate: self)
                    vc.applyTheme(withContainerScheme: self.scheme)
                    self.rootViewController.pushViewController(vc, animated: animated)
                    self.navigationControllerObserver.observePopTransition(of: vc, delegate: self)
                    self.hasPushedViewController = true
                    if UIDevice.current.userInterfaceIdiom == .pad {
                        vc.navigationItem.leftBarButtonItem = UIBarButtonItem(
                            image: UIImage(systemName: "xmark"), style: .plain,
                            target: self, action: #selector(self.dismiss(_:))
                        )
                    }
                }
            } else if let urlToLoad = self.urlToLoad {
                if urlToLoad.isFileURL || DataConnectionUtilities.shouldFetchAvatars() || ImageCache.default.isCached(forKey: urlToLoad.absoluteString) {
                    self.loadURL(animated: animated)
                } else {
                    let vc = AskToDownloadViewController(url: urlToLoad, delegate: self)
                    vc.applyTheme(withContainerScheme: self.scheme)
                    self.rootViewController.pushViewController(vc, animated: animated)
                    self.navigationControllerObserver.observePopTransition(of: vc, delegate: self)
                    self.hasPushedViewController = true
                    if UIDevice.current.userInterfaceIdiom == .pad {
                        vc.navigationItem.leftBarButtonItem = UIBarButtonItem(
                            image: UIImage(systemName: "xmark"), style: .plain,
                            target: self, action: #selector(self.dismiss(_:))
                        )
                    }
                }
            }
        }
    }

    @objc public func setAttachment(attachment: AttachmentModel) {
        self.attachment = attachment
        onMain {
            if self.hasPushedViewController {
                self.ignoreNextDelegateCall = true
                FadeTransitionSegue.addFadeTransition(to: self.rootViewController.view)
                self.rootViewController.popViewController(animated: false)
            }
            self.start(false)
        }
    }

    private func loadURL(animated: Bool = false) {
        guard let urlToLoad = self.urlToLoad, let contentType = self.contentType else { return }
        onMain {
            if contentType.isImageContentType {
                let imageVC = ImageAttachmentViewController(url: urlToLoad)
                imageVC.view.backgroundColor = .black
                self.rootViewController.pushViewController(imageVC, animated: animated)
                self.navigationControllerObserver.observePopTransition(of: imageVC, delegate: self)
                self.hasPushedViewController = true
                self.imageViewController = imageVC
                if UIDevice.current.userInterfaceIdiom == .pad {
                    imageVC.navigationItem.leftBarButtonItem = UIBarButtonItem(
                        image: UIImage(systemName: "xmark"), style: .plain,
                        target: self, action: #selector(self.dismiss(_:))
                    )
                }
            } else if contentType.isVideoContentType || contentType.isAudioContentType {
                self.playAudioVideo()
            }
        }
    }

    func downloadApproved() {
        MageLogger.misc.debug("Download the attachment")
        onMain {
            FadeTransitionSegue.addFadeTransition(to: self.rootViewController.view)
            self.ignoreNextDelegateCall = true
            self.rootViewController.popViewController(animated: false)
            if self.attachment != nil {
                self.showAttachment()
            } else {
                self.loadURL()
            }
        }
    }

    private func showAttachment(animated: Bool = false) {
        guard let attachment = self.attachment, let contentType = attachment.contentType else { return }
        onMain {
            if contentType.isImageContentType {
                let imageVC = ImageAttachmentViewController(attachment: attachment)
                imageVC.view.backgroundColor = .black
                self.rootViewController.pushViewController(imageVC, animated: animated)
                self.navigationControllerObserver.observePopTransition(of: imageVC, delegate: self)
                self.hasPushedViewController = true
                self.imageViewController = imageVC
                if UIDevice.current.userInterfaceIdiom == .pad {
                    imageVC.navigationItem.leftBarButtonItem = UIBarButtonItem(
                        image: UIImage(systemName: "xmark"), style: .plain,
                        target: self, action: #selector(self.dismiss(_:))
                    )
                }
                return
            }

            if contentType.isVideoContentType || contentType.isAudioContentType {
                self.playAudioVideo()
                return
            }

            // Other file types
            var url: URL?
            if let localURL = self.healedLocalURL(for: attachment) {
                url = localURL
            } else if let s = attachment.url, let remote = URL(string: s) {
                url = AccessTokenURL.tokenized(remote)
            }

            if let url {
                self.mediaPreviewController = MediaPreviewController(
                    fileName: attachment.name ?? "file",
                    mediaTitle: attachment.name ?? "file",
                    data: nil,
                    url: url,
                    mediaLoaderDelegate: self,
                    scheme: self.scheme
                )
                self.rootViewController.pushViewController(self.mediaPreviewController!, animated: true)
                if UIDevice.current.userInterfaceIdiom == .pad {
                    self.mediaPreviewController?.navigationItem.leftBarButtonItem = UIBarButtonItem(
                        image: UIImage(systemName: "xmark"), style: .plain,
                        target: self, action: #selector(self.dismiss(_:))
                    )
                }
            } else {
                MDCSnackbarManager.default.show(MDCSnackbarMessage(text: "Unable to open attachment \(attachment.name ?? "file")"))
            }
        }
    }

    private func playAudioVideo() {
        var name = "file"
        if let attachment = self.attachment {
            name = attachment.name ?? "file"
            if let localURL = healedLocalURL(for: attachment) {
                MageLogger.misc.debug("Playing locally: \(localURL.path)")
                self.urlToLoad = localURL
            } else if let s = attachment.url, let remote = URL(string: s) {
                MageLogger.misc.debug("Playing from link \(remote.absoluteString)")
                self.urlToLoad = AccessTokenURL.tokenized(remote)
            }
        }

        guard let urlToLoad = self.urlToLoad else { return }

        onMain {
            self.mediaPreviewController = MediaPreviewController(
                fileName: name, mediaTitle: name, data: nil, url: urlToLoad,
                mediaLoaderDelegate: self, scheme: self.scheme
            )
            self.rootViewController.pushViewController(self.mediaPreviewController!, animated: true)
            if UIDevice.current.userInterfaceIdiom == .pad {
                self.mediaPreviewController?.navigationItem.leftBarButtonItem = UIBarButtonItem(
                    image: UIImage(systemName: "xmark"), style: .plain,
                    target: self, action: #selector(self.dismiss(_:))
                )
            }
        }
    }

    // MARK: - NavigationControllerObserverDelegate (must remain nonisolated)

    func navigationControllerObserver(_ observer: NavigationControllerObserver, didObservePopTransitionFor viewController: UIViewController) {
        onMain {
            if !self.ignoreNextDelegateCall {
                self.delegate?.doneViewing(coordinator: self)
            }
            self.ignoreNextDelegateCall = false
        }
    }

    @objc func dismiss(_ sender: UIBarButtonItem) {
        onMain {
            self.rootViewController.dismiss(animated: true, completion: nil)
        }
    }
}

// MARK: - AVAssetResourceLoaderDelegate
extension AttachmentViewCoordinator: AVAssetResourceLoaderDelegate {}

// MARK: - MediaLoaderDelegate (must remain nonisolated)
extension AttachmentViewCoordinator: MediaLoaderDelegate {
    func mediaLoadComplete(filePath: String, newFile: Bool) {
        MageLogger.misc.debug("BBB: Media load complete")
        guard newFile else { return }

        // Update in-memory model for immediate UI
        let relative = AttachmentPath.stripToDocumentsRelative(filePath)
        self.attachment?.localPath = relative

        // Persist to Core Data (normalization happens inside)
        attachmentRepository.saveLocalPath(attachmentUri: attachment?.attachmentUri, localPath: filePath)
        // or: localSD.saveLocalPath(attachmentUri: attachment?.attachmentUri, localPath: filePath)
    }
}
